File: src/main/java/org/muieer/java/leetcode/LeetCode392.java
package org.muieer.java.leetcode;

/*
* 判断子序列：https://leetcode.cn/problems/is-subsequence/
* 时间复杂度 O(n)
* 空间复杂度 O(1)
* 思路：双指针
* */
public class LeetCode392 {

    public static void main(String[] args) {
        var demo = new LeetCode392();
        System.out.println(demo.isSubsequence("ac", "ax"));
    }

    public boolean isSubsequence(String s, String t) {

        if (s.length() == 0) return true;
        if (t.length() == 0) return false;

        int ls = 0, rs = s.length() - 1;
        int lt = 0, rt = t.length() - 1;

        while(ls <= rs) {

            // 找到 s 字符串左指针在 t 中对应位置
            while (lt <= rt && t.charAt(lt) != s.charAt(ls)) {
                lt++;
            }
            // 找到了
            if (lt <= rt) {
                ls++;
                lt++;
            }

            while (lt <= rt && t.charAt(rt) != s.charAt(rs)) {
                rt--;
            }
            // 找到了
            if (lt <= rt) {
                rs--;
                rt--;
            }

            if (lt > rt) break;
        }

        return ls > rs;
    }
}

File: src/main/scala/org/muieer/scala/leetcode/LeetCode392.scala
package org.muieer.scala.leetcode

object LeetCode392 {

  def main(args: Array[String]): Unit = {
    println(isSubsequence("aaaaaa", "bbaaaa"))
  }

  def isSubsequence(s: String, t: String): Boolean = {
    if (s.isEmpty) return true
    if (t.isEmpty) return false

    var si = 0
    var ti = 0

    while (si < s.length && ti < t.length) {
      // 依次匹配每个待匹配的字符
      while (ti < t.length && t(ti) != s(si)) {
        ti = ti + 1
      }
      // 匹配成功一个字符
      if (ti != t.length) {
        si = si + 1 // 指针右移
        ti = ti + 1 // 指针右移
      }
    }

    si == s.length
  }


}
